### ISSUE TITLE
Decrease in property value of 'limit' in 'FluxFlatMap#FlatMapInner'

### ISSUE THREAD

**Issue by phantasmicmeans** [reactor/reactor-core#2586]
## Motivation
It's a little similar to issue #2571, but there's definitely a difference, so I made it a new issue. 

As we already know, 'flatMap' has two parameters: 'concurrency' and 'prefetch'.
- concurrency: request multiple values from the source 
- prefetch: request on each inner publisher 

### Code
```java
//source code 
Flux<Integer> flux = Flux.fromStream(boxed1);
Function<Integer, Publisher<Integer>> function = v -> Flux.fromStream(IntStream.range(1, 8).boxed()).log();
flux.log()
   .flatMap(function, 1, 8) // concurrency, prefetch 
   .subscribe();
```

Of course, I thought the value 'n' of `request(n)` for inner publisher would be fixed at 8. But when I ran it, I got the following results.

`request(8) -> request(6) -> request(8) -> requeust(6) ...`

As you can see, there seems to be one more request. (request(6))

### Log 

```bash
// console 
INFO 4783 --- [    Test worker] reactor.Flux.Stream.1  : | onSubscribe([Synchronous Fuseable] FluxIterable.IterableSubscription)
INFO 4783 --- [    Test worker] reactor.Flux.Stream.1  : | request(1)
INFO 4783 --- [    Test worker] reactor.Flux.Stream.1  : | onNext(1)
INFO 4783 --- [    Test worker] reactor.Flux.Stream.2  : | onSubscribe([Synchronous Fuseable] FluxIterable.IterableSubscription)
INFO 4783 --- [    Test worker] reactor.Flux.Stream.2  : | request(8)
INFO 4783 --- [    Test worker] reactor.Flux.Stream.2  : | onNext(1)
INFO 4783 --- [    Test worker] reactor.Flux.Stream.2  : | onNext(2)
INFO 4783 --- [    Test worker] reactor.Flux.Stream.2  : | onNext(3)
INFO 4783 --- [    Test worker] reactor.Flux.Stream.2  : | onNext(4)
INFO 4783 --- [    Test worker] reactor.Flux.Stream.2  : | onNext(5)
INFO 4783 --- [    Test worker] reactor.Flux.Stream.2  : | onNext(6)
INFO 4783 --- [    Test worker] reactor.Flux.Stream.2  : | request(6)
INFO 4783 --- [    Test worker] reactor.Flux.Stream.2  : | onNext(7)
INFO 4783 --- [    Test worker] reactor.Flux.Stream.2  : | onComplete()
INFO 4783 --- [    Test worker] reactor.Flux.Stream.1  : | request(1)
INFO 4783 --- [    Test worker] reactor.Flux.Stream.1  : | onNext(2)
INFO 4783 --- [    Test worker] reactor.Flux.Stream.3  : | onSubscribe([Synchronous Fuseable] FluxIterable.IterableSubscription)
INFO 4783 --- [    Test worker] reactor.Flux.Stream.3  : | request(8)
INFO 4783 --- [    Test worker] reactor.Flux.Stream.3  : | onNext(1)
INFO 4783 --- [    Test worker] reactor.Flux.Stream.3  : | onNext(2)
INFO 4783 --- [    Test worker] reactor.Flux.Stream.3  : | onNext(3)
...
```

I followed the source code and found that 'FlatMapInner' and  'LambdaSubscriber(actual)' instances were used.

### FluxFlatMap#FlatMapInner, Operators.unboundedOrLimit
```java
// FluxFlatMap#FlatMapInner
static final class FlatMapInner<R> implements InnerConsumer<R>, Subscription {
...

   FlatMapInner(FlatMapMain<?, R> parent, int prefetch) {
      this.parent = parent;
      this.prefetch = prefetch;
   // this.limit = prefetch >> 2;
      this.limit = Operators.unboundedOrLimit(prefetch); // here 

...

   @Override
   public void request(long n) {
      long p = produced + n;
      if (p >= limit) {  // here, request more when limit count arrived
         produced = 0L;
         s.request(p); // request
      }
      else {
         produced = p;
      }
   }

...

}

// Operators.unboundedOrLimit
static int unboundedOrLimit(int prefetch) {
   return prefetch == Integer.MAX_VALUE ? Integer.MAX_VALUE : (prefetch - (prefetch >>	2));
}
```

It was not until I saw the method "unboundedOrLimit" that I could see why `n` decreased and one more request has occured.

It seems that developers may be confused by unexpected request.

## Desired solution
Although there is no final keyword, the `prefetch` value is an important parameter **controlled by the user**. In addition, few people would think that this value 'n' would be changed 'prefetch' to a 'reduced limit' because this value is part of the `user's own definition of backpressure.`

Is the use of reduced prefetch as 'limit' to improve performance?

Since it's a 'backpressure' defined by the user, why don't we fix the size of limit (prefetch) and remove the additional request?

Or why don't we request when the 'prefetch' size is larger than a certain size?

**Comment by simonbasle** [comment]
This replenishing limit pattern is a way to improve performance, yes. it keeps the inner publisher warmed up in case of incremental request (vs unbounded prefetch) and allows to absorb their latency a bit. This historical pattern was put in place because in a majority of cases, requesting a cold source one by one is inefficient.

See David Karnok's blog [flatMap part 2](https://akarnokd.blogspot.com/2016/03/flatmap-part-2.html) for excellent insight on the design and implementation of `flatMap`. Here is an extract:

> This point can be anywhere between 1 and the prefetch amount and generally, the value really depends on how the source emits values. Sources may work better at any point in this range. Unfortunately, libraries can't help too much setting these individually and any adaptive logic adds so much overhead that likely negates any benefit from it. Therefore, RxJava chose to re-request after half of the prefetch amount has been emitted (and lately, I tend to use 75% of the prefetch for it).

The limit could be set at `prefetch`, but it has more potential for "wasting" some of these elements. It could be `1`, but we'd be back at the same point of compounded latency.

Also keep in mind the replenishment should NOT continue more than once after the current downstream request is entirely served (ie. is at zero).

`flatMap` is a pretty complex operator. It is one of the most critical components in terms of performance, and has to juggle multiple sources in a way that maintains consistency with the spec while interleaving data into a single subscriber all the while following the RS spec from an outer perspective...

`concurrency` is a straightforward lever, while `prefetch` is a bit more subtle.

**Comment by phantasmicmeans** [comment]
As you said, `flatmap` is a complicated operator. 
Thank you for the nice explanation.
